File: py/1128.py
from collections import defaultdict


class Graph:
    def __init__(self, vertices) -> None:
        self.vertices = vertices
        self.graph = defaultdict(list)

    def add_edge(self, f, t):
        self.graph[f].append(t)

    def dfs(self, vertex, visited):
        visited[vertex] = True
        for i in self.graph[vertex]:
            if not visited[i]:
                self.dfs(i, visited)

    def get_transposed(self):
        graph = Graph(self.vertices)
        for i in self.graph:
            for j in self.graph[i]:
                graph.add_edge(j, i)
        return graph

    def is_strongly_connected(self):
        visited = [False] * self.vertices
        self.dfs(0, visited)
        for i in visited:
            if not i:
                return False
        graph_transp = self.get_transposed()
        visited = [False] * self.vertices
        graph_transp.dfs(0, visited)
        for i in visited:
            if not i:
                return False
        return True


def main() -> None:
    while True:
        n, m = [int(i) for i in input().split()]
        graph = Graph(n)
        if n == 0 and m == 0:
            break
        for _ in range(m):
            v, w, p = [int(i) for i in input().split()]
            v -= 1
            w -= 1
            graph.add_edge(v, w)
            if p == 2:
                graph.add_edge(w, v)
        print(int(graph.is_strongly_connected()))


if __name__ == '__main__':
    main()

File: cpp/1128.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Graph {
    public:
        int vertices;
        unordered_map<int, vector<int>> graph;

    void add_edge(int f, int t) {
        graph[f].push_back(t);
    }

    void dfs(int vertex, bool visited[]) {
        visited[vertex] = true;
        for (auto i: graph[vertex]) {
            if (!visited[i]) {
                dfs(i, visited);
            }
        }
    }

    Graph get_transposed() {
        Graph new_graph;
        new_graph.vertices = vertices;
        for (auto i: graph) {
            for (auto j: i.second) {
                new_graph.add_edge(j, i.first);
            }
        }
        return new_graph;
    }

    bool is_strongly_connected() {
        bool visited[vertices] = {false};
        dfs(0, visited);
        for (auto i: visited) {
            if (!i) {
                return false;
            }
        }
        bool t_visited[vertices] = {false};
        Graph new_graph;
        new_graph = get_transposed();
        new_graph.dfs(0, t_visited);
        for (auto i: t_visited) {
            if (!i) {
                return false;
            }
        }
        return true;
    }
};

int main()
{
    int n, m, v, w, p;
    while (true) {
        cin >> n >> m;
        Graph gr;
        gr.vertices = n;
        if (n == 0 && m == 0) {
            break;
        }
        for (m; m > 0; m--) {
            cin >> v >> w >> p;
            v--;
            w--;
            gr.add_edge(v, w);
            if (p == 2) {
                gr.add_edge(w, v);
            }
        }
        cout << gr.is_strongly_connected() << '\n';
    }
    return 0;
}